src/providers.py

```diff
@@ -0,0 +1,116 @@
+from io import BytesIO, StringIO
+
+import enry
+from dulwich import repo, patch
+from tree_sitter import Parser, Tree
+from unidiff import PatchSet, UnidiffParseError, PatchedFile
+
+from src.constants import languages
+from src.helpers import FileSystemHelper
+from src.models import Change, Repository, Commit, File
+from src.services import GitService
+from src.setup import build_tree_sitter
+
+
+class LanguagesProvider:
+    def __init__(self, repository_path: str) -> None:
+        self.repository_path = repository_path
+        self.languages_counts = {}
+        self.files_with_languages = []
+
+    def add_language(self, language: str) -> None:
+        if language not in self.languages_counts.keys():
+            self.languages_counts[language] = 1
+        else:
+            self.languages_counts[language] += 1
+
+    def add_file(self, file: File, language: str) -> None:
+        file_with_language = File(
+            name=file.name,
+            path=file.path,
+            language=language)
+        self.files_with_languages.append(file_with_language)
+
+    def get_language(self, file_name: str, content: str) -> str:
+        language = enry.get_language(file_name, str.encode(content))
+        if language == '':
+            return 'Other'
+        return language
+
+    def calc_languages(self) -> None:
+        for file in FileSystemHelper.get_files(self.repository_path):
+            with open(file.name_with_path) as f:
+                try:
+                    language = self.get_language(file.name, f.read())
+                    self.add_language(language)
+                    self.add_file(file, language)
+                except UnicodeDecodeError:
+                    pass
+
+    def get_language_ratio(self) -> {str: float}:
+        return {k: self.languages_counts[k] / sum(self.languages_counts[k] for k in self.languages_counts) for k in
```

Comment:
* let's use meaningful names `k` -> `language`
* you can calculate sum of language counts only once to optimize it a bit